File: server/src/main/java/de/zalando/zally/dto/ViolationsCounter.java
package de.zalando.zally.dto;

import de.zalando.zally.rule.Result;
import de.zalando.zally.rule.api.Severity;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ViolationsCounter {
    private final Map<Severity, Integer> counters;

    public ViolationsCounter(List<Result> violations) {
        counters = violations.
            stream().
            collect(Collectors.groupingBy(Result::getViolationType, Collectors.summingInt(x -> 1)));
    }

    public Integer getCounter(Severity violationType) {
        return counters.getOrDefault(violationType, 0);
    }
}

File: server/src/main/java/de/zalando/zally/dto/ViolationsCounter.kt
package de.zalando.zally.dto

import de.zalando.zally.core.Result
import de.zalando.zally.rule.api.Severity

class ViolationsCounter(violations: List<Result>) {

    private val counters = violations.groupingBy { it.violationType }.eachCount()

    operator fun get(severity: Severity): Int = counters[severity] ?: 0
    fun getCounter(violationType: Severity): Int = counters[violationType] ?: 0
}
